Sources/SwiftFormatPrettyPrint/PrettyPrint.swift

```diff
@@ -43,66 +31,129 @@ public class PrettyPrinter {
+  /// delimiters and whether or not the current line is a continuation line.
+  private var currentIndentation: [Indent] {
+    var totalIndentation = indentationStack
+    if currentLineIsContinuation {
+      totalIndentation.append(configuration.indentation)
+    }
+    return totalIndentation
   }
 
   /// Creates a new PrettyPrinter with the provided formatting configuration.
   ///
   /// - Parameters:
   ///   - context: The formatter context.
   ///   - node: The node to be pretty printed.
-  public init(context: Context, node: Syntax, isDebugMode: Bool, printTokenStream: Bool) {
+  public init(context: Context, node: Syntax, printTokenStream: Bool) {
     self.context = context
     let configuration = context.configuration
     self.tokens = node.makeTokenStream(configuration: configuration)
     self.maxLineLength = configuration.lineLength
-    self.isDebugMode = isDebugMode
     self.spaceRemaining = self.maxLineLength
     self.printTokenStream = printTokenStream
   }
 
-  /// Append the input string to the output buffer
-  func write<S: StringProtocol>(_ str: S) {
+  /// Append the given string to the output buffer.
+  ///
+  /// No further processing is performed on the string.
+  private func writeRaw<S: StringProtocol>(_ str: S) {
     outputBuffer.append(String(str))
   }
 
+  /// Writes the given number of newlines to the output stream.
+  ///
+  /// - Parameters:
+  ///   - count: The number of newlines to write.
+  ///   - discretionary: Indicates whether the newlines are user-entered discretionary newlines.
+  ///     Discretionary newlines are always printed (up through the maximum allowed number provided
+  ///     to the printer at initialization time). Non-discretionary newlines are only printed if
+  ///     discretionary newlines have already not been printed yet.
+  private func writeNewlines(_ count: Int, discretionary: Bool) {
+    // We add 1 because it takes 2 newlines to create a blank line.
+    let maximumNewlines = configuration.maximumBlankLines + 1
+    let numberToPrint: Int
+    if consecutiveNewlineCount + count <= maximumNewlines {
+      numberToPrint = count - consecutiveNewlineCount
+    } else {
+      numberToPrint = maximumNewlines - consecutiveNewlineCount
+    }
+
+    guard (discretionary && numberToPrint > 0) || consecutiveNewlineCount == 0 else { return }
+
+    writeRaw(String(repeating: "\n", count: numberToPrint))
+    lineNumber += numberToPrint
+    isAtStartOfLine = true
+    consecutiveNewlineCount += numberToPrint
+    pendingSpaces = 0
+  }
+
+  /// Request that the given number of spaces be printed out before the next text token.
+  ///
+  /// Spaces are printed only when the next text token is printed in order to prevent us from
+  /// printing lines that are only whitespace or have trailing whitespace.
+  private func enqueueSpaces(_ count: Int) {
+    pendingSpaces += count
+    spaceRemaining -= count
```

Comment:
When we check the next token after the spaces to see if it fits, we need to check it against the space remaining that takes into account the enqueued spaces, otherwise it might think it has more room than it actually does. If it doesn't fit, then we'll move to the next line and just throw out those pending spaces. If we wait until we print them, it's too late and we'll end up with trailing whitespace.